src/firebase_functions/pubsub.py

```diff
@@ -20,69 +24,110 @@
 
 T = TypeVar("T")
 
-Request = flask.Request
-Response = flask.Response
-
 
 @dataclass(frozen=True)
 class CloudEvent(Generic[T]):
+    id: str
+    datacontenttype: str
     specversion: str
     source: str
-    subject: str
     type: str
     time: dt.datetime
     data: T
 
 
 @dataclass(frozen=True)
 class Message(Generic[T]):
+    """
+    Wrapper around a Pub/Sub message.
+    """
     message_id: str
     publish_time: dt.datetime
-    data: str
-    attributes: dict[str, str]
-    ordering_key: str
+    data: Optional[str] = None
+    attributes: Optional[dict[str, str]] = None
+    ordering_key: Optional[str] = None
 
     @property
-    def json(self) -> Dict[str, Any]:
-        return {
+    def json(self) -> Optional[T]:
+        try:
+            if self.data is not None:
+                return json.loads(base64.b64decode(self.data).decode("utf-8"))
+            else:
+                return None
+        except Exception as e:
+            raise Exception(
+                f"Unable to parse Pub/Sub message data as JSON: {e}") from e
+
+    def asdict(self) -> dict[str, Any]:
+        dict_message: dict[str, Any] = {
             "message_id": self.message_id,
-            "publish_time": self.publish_time,
             "data": self.data,
-            "attributes": self.attributes,
-            "ordering_key": self.ordering_key,
+            "publish_time": self.publish_time,
         }
 
+        if self.attributes is not None:
+            dict_message["attributes"] = self.attributes
+        if self.ordering_key is not None:
+            dict_message["ordering_key"] = self.ordering_key
 
-CloudEventMessage = CloudEvent[Message[str]]
+        return dict_message
 
 
-class MessagePublishedData(TypedDict):
-    message: object
+@dataclass(frozen=True)
+class MessagePublishedData(Generic[T]):
+    message: Message[T]
     subscription: str
 
 
 def pubsub_wrap_handler(
-    func: Callable[[CloudEventMessage], None],
-    raw: dict[str, Any],
-) -> Response:
-    event: CloudEvent[Message[str]] = CloudEvent(
-        **raw,
-        time=dt.datetime.fromisoformat(raw["time"]),
-        data=Message(
-            **raw["data"]["message"],
-            publish_time=dt.datetime.fromisoformat(
-                raw["data"]["message"]["publish_time"]
-            ),
+    func: Callable[[CloudEvent[MessagePublishedData[T]]], None],
+    raw: ce.CloudEvent,
+) -> flask.Response:
+    if isinstance(raw, ce.CloudEvent):
+        event_dict = {**raw}
+    else:
+        event_dict = raw
```

Comment:
What is the `else` here for, raw is typed as a CloudEvent which indicates to me it can only be a cloud event, if that's not the case then we should update the types, my guess is it's always a dict but either way if CloudEvent or Dict then this can be simplified to:

```suggestion
    raw: Union[ce.CloudEvent, dict],
) -> flask.Response:
    event_dict = {**raw}
```

Reply:
This is outdated now since the force push, but comment still applies

Reply:
Code coverage also seems to indicate that it's always a `dict` 🤔 so maybe it is always dict?
![image](https://user-images.githubusercontent.com/5347038/183651557-8342a2d7-6c1c-4af7-808c-0ab0ded9bdac.png)

Reply:
in tests it comes as a dict, but on cloud it comes as a cloudevent, so I'm not sure how to solve it for tests

Reply:
can you add a TODO on the `else:` statement where it handles the dict, and just say something like 'used for tests only, refactor tests to use CloudEvent instead of dict' and then we can look at it on a later date